Add createworkflow failure proto #minor

jerempy opened the pull request:
# TL;DR
Adds createworkflow failure proto. This is for more concrete return values when a workflow fails to get created.
For context: https://github.com/flyteorg/flyteadmin/pull/487#issuecomment-1281212131

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

https://github.com/flyteorg/flyte/issues/2522

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

jerempy commented:
@katrogan Would you mind taking a look at this? If it looks good to you, would you mind helping me compile the protobufs? This is out of my wheelhouse unfortunately 😞

katrogan commented on protos/flyteidl/core/workflow.proto:
```diff
@@ -205,3 +205,21 @@ message TaskNodeOverrides {
+
+// The workflow id is already used and the structure is different
+message WorkflowErrorExistsDifferentStructure {
+    string id = 1;
```
let's use the workflow identifier, like here:  https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/workflow.proto#L16

katrogan commented on protos/flyteidl/core/workflow.proto:
```diff
@@ -205,3 +205,21 @@ message TaskNodeOverrides {
+}
+
+// The workflow id is already used with an identical sctructure
+message WorkflowErrorExistsIdenticalStructure {
```
I wonder if we need this message since we return a different gRPC status code (AlreadyExists) for this scenario

katrogan replied:
although perhaps it would be nice to keep this to include the existing workflow id 👍

katrogan reviewed (commented):
hey @jerempy thanks for opening this! can you try [these instructions](https://github.com/flyteorg/flyteidl#tooling-for-flyteidl) for generating proto code and let me know if that works for you?

jsmith-chi commented:
> hey @jerempy thanks for opening this! can you try [these instructions](https://github.com/flyteorg/flyteidl#tooling-for-flyteidl) for generating proto code and let me know if that works for you?

Hey there, I did try that and had difficulties. It looks dated since there is no "download_tooling" in the makefile. I tried looking at different things in the make file and couldn't figure out what I was missing. Any other ideas?

jerempy commented:
> hey @jerempy thanks for opening this! can you try [these instructions](https://github.com/flyteorg/flyteidl#tooling-for-flyteidl) for generating proto code and let me know if that works for you?

Hey there, I did try that and had difficulties. It looks dated since there is no "download_tooling" in the makefile. I tried looking at different things in the make file and couldn't figure out what I was missing. Any other ideas?

katrogan commented:
ah @jerempy the top-level [Makefile](https://github.com/flyteorg/flyteidl/blob/b24a555934138df9fb3d710fbbf77084cc395821/Makefile#L4) should import the boilerplate one with the target: https://github.com/flyteorg/flyteidl/blob/b24a555934138df9fb3d710fbbf77084cc395821/boilerplate/flyte/golang_test_targets/Makefile#L7

jerempy commented:
@katrogan thanks for all the input! I updated the type for the id - good spotting that!
Thanks for the direction on getting the compiler to work, but I just seem to not be having any luck. Maybe its cuz im on windows 10. I tried through my cmd prompt. I tried through wsl, and I tried to git bash. 

When trying to run `make` command in wsl I would get this error:

```
Makefile:17: warning: overriding recipe for target 'generate'
boilerplate/flyte/golang_test_targets/Makefile:13: warning: ignoring old recipe for target 'generate'
Makefile:30: warning: overriding recipe for target 'test_unit'
boilerplate/flyte/golang_test_targets/Makefile:38: warning: ignoring old recipe for target 'test_unit'
make: boilerplate/flyte/golang_test_targets/download_tooling.sh: Command not found
make: *** [boilerplate/flyte/golang_test_targets/Makefile:9: download_tooling] Error 127
```

I tried running the .sh scripts directy and that didn't turn out so good. Wanted to delete 1200 files as a result 😨 
So for now I just changed the types on the id and I think I'm a bit out of luck on this one. unfortunately. I might try again tomorrow

katrogan commented:
hey @jerempy do you mind filing an issue dumping what you ran into?

I can help you with the generation for now. But in the meantime, could I ask you to move the new proto message definitions to the admin workflow proto def: https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/workflow.proto rather than the core one you've defined them in now?

jerempy commented:
@katrogan ok great. thanks much for your help! and good spottin on the location - I moved it. Thanks! 🙏

katrogan commented:
@EngHabu since you also reviewed the corresponding admin [PR](https://github.com/flyteorg/flyteadmin/pull/487), mind reviewing this one too?

EngHabu commented on protos/flyteidl/admin/workflow.proto:
```diff
@@ -65,3 +65,21 @@ message WorkflowClosure {
+}
+
+// When a CreateWorkflowRequest failes due to matching id
+message CreateWorkflowFailureReason {
```
Should we do the same for tasks?

jerempy replied:
What if we do that in a seperate issue/ pr series? I can make the issue for it as a follow up after finishing this one if you want.

EngHabu replied:
That would be awesome!

EngHabu commented on protos/flyteidl/admin/workflow.proto:
```diff
@@ -65,3 +65,21 @@ message WorkflowClosure {
+
+// The workflow id is already used and the structure is different
+message WorkflowErrorExistsDifferentStructure {
+    core.Identifier id = 1;
```
Should we include the structure of the other workflow?

jerempy replied:
Current plan was to return the id of the workflow to the sdk client and then to get the workflow and as part of the exception handling to then compare the diff with the initial request

EngHabu reviewed (commented):
Thank you for taking on that work! Mind linking to an issue too?

jerempy commented:
@EngHabu thanks for reviewing! I updated the top comment to Include the issue. And thanks @katrogan for compiling for me!

jerempy commented:
@EngHabu @katrogan if possible, think we could merge this today? After this one is done then I still need to re-do the flyeadmin PR and then the flytekit PR to get this issue resolved and I only got 7 days left until the end of the month 🙏
